_posts/2019-08-23-unet/unet.Rmd

```diff
@@ -0,0 +1,437 @@
+    # decode_jpeg yields a 3d tensor of shape (1280, 1918, 3)
+    img = tf$image$decode_jpeg(tf$io$read_file(.x$img)),
+    # decode_gif yields a 4d tensor of shape (1, 1280, 1918, 3),
+    # so we remove the unneeded batch dimension and all but one 
+    # of the 3 (identical) channels
+    mask = tf$image$decode_gif(tf$io$read_file(.x$mask))[1,,,][,,1,drop=FALSE]
+  ))
+```
+
+While constructing a preprocessing pipeline, it's very useful to check intermediate results.
+It's easy to do using `reticulate::as_iterator` on the dataset:
+
+```{r}
+example <- training_dataset %>% as_iterator() %>% iter_next()
+example
+```
+
+```
+$img
+tf.Tensor(
+[[[243 244 239]
+  [243 244 239]
+  [243 244 239]
+  ...
+ ...
+  ...
+  [175 179 178]
+  [175 179 178]
+  [175 179 178]]], shape=(1280, 1918, 3), dtype=uint8)
+
+$mask
+tf.Tensor(
+[[[0]
+  [0]
+  [0]
+  ...
+ ...
+  ...
+  [0]
+  [0]
+  [0]]], shape=(1280, 1918, 1), dtype=uint8)
+
+```
+
+While the `uint8` datatype makes RGB values easy to read for humans, the network is going to expect floating point numbers. The following code converts its input and additionally, scales values to the interval [0,1):
+
+
+```{r}
+training_dataset <- training_dataset %>% 
+  dataset_map(~.x %>% list_modify(
+    img = tf$image$convert_image_dtype(.x$img, dtype = tf$float32),
+    mask = tf$image$convert_image_dtype(.x$mask, dtype = tf$float32)
+  ))
+```
+
+To reduce computational cost, we resize the images to size `128x128`. This will change the aspect ratio and thus, distort the images, but is not a problem with the given dataset.
+
+```{r}
+training_dataset <- training_dataset %>% 
+  dataset_map(~.x %>% list_modify(
+    img = tf$image$resize(.x$img, size = shape(128, 128)),
+    mask = tf$image$resize(.x$mask, size = shape(128, 128))
+  ))
+```
+
+Now, it's well known that in deep learning, data augmentation is paramount. For segmentation, there's one thing to consider, which is whether a transformation needs to be applied to the mask as well -- this would be the case for e.g. rotations, or flipping. Here, results will be good enough applying just transformations that preserve positions:
+
+```{r}
+random_bsh <- function(img) {
+  img %>% 
+    tf$image$random_brightness(max_delta = 0.3) %>% 
+    tf$image$random_contrast(lower = 0.5, upper = 0.7) %>% 
+    tf$image$random_saturation(lower = 0.5, upper = 0.7) %>% 
+    # make sure we still are between 0 and 1
+    tf$clip_by_value(0, 1) 
+}
+
+training_dataset <- training_dataset %>% 
+  dataset_map(~.x %>% list_modify(
+    img = random_bsh(.x$img)
+  ))
+
+```
+
+Again, we can use `as_iterator` to see what these transformations do to our images:
+
+```{r}
+example <- training_dataset %>% as_iterator() %>% iter_next()
+example$img %>% as.array() %>% as.raster() %>% plot()
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = ""}
+knitr::include_graphics("images/transformed.png")
+```
+
+Now here's the complete preprocessing pipeline.
```

Comment:
```suggestion
Here's the complete preprocessing pipeline.
```

Reply:
Since you start quite a few lines with "Now," I might recommend mixing it up by removing it in a few places.

Reply:
Totally. Actually I removed that and put it back again :-(
My normal procedure is "remove as many fill words as possible when proofreading".
Sometimes I'm wondering if it might sound too "hard", but if you say remove, it's remove :-)